File: scripts/merge-batches.ts
/**
 * Script to merge processed batch files back into the original quiz
 */

import * as fs from 'fs';
import * as path from 'path';

interface QuizOption {
  optionId: string;
  optionText: string;
}

interface ProcessedQuestion {
  chapterIndex: number;
  questionIndex: number;
  questionId: string;
  questionText: string;
  options: QuizOption[];
  correctOptionIds: string[];
  explanationText: string;
}

interface ProcessedBatch {
  batchIndex: number;
  questions: ProcessedQuestion[];
}

interface Question {
  questionId: string;
  questionText: string;
  options: QuizOption[];
  correctOptionIds: string[];
  explanationText: string;
  type: string;
  status: string;
  timesAnsweredCorrectly: number;
  timesAnsweredIncorrectly: number;
  historyOfIncorrectSelections: string[];
  lastSelectedOptionId: string | null;
  lastAttemptedAt: string | null;
  srsLevel: number;
  nextReviewAt: string | null;
  shownIncorrectOptionIds: string[];
}

interface Chapter {
  id: string;
  name: string;
  questions: Question[];
  totalQuestions: number;
  answeredQuestions: number;
  correctAnswers: number;
  isCompleted: boolean;
}

interface Quiz {
  name: string;
  description: string;
  chapters: Chapter[];
}

// Read the original quiz
const quizPath = path.join(__dirname, '../public/default-quiz.json');
const quiz: Quiz = JSON.parse(fs.readFileSync(quizPath, 'utf-8'));

// Read all processed batch files
const batchDir = path.join(__dirname, '../temp-batches');
const processedDir = path.join(__dirname, '../temp-batches/processed');

if (!fs.existsSync(processedDir)) {
  console.error('No processed directory found. Run subagents first.');
  process.exit(1);
}

const processedFiles = fs.readdirSync(processedDir).filter((f) => f.endsWith('.json'));
console.log(`Found ${processedFiles.length} processed batch files`);

let updatedCount = 0;

processedFiles.forEach((file) => {
  const filePath = path.join(processedDir, file);
  const batch: ProcessedBatch = JSON.parse(fs.readFileSync(filePath, 'utf-8'));

  batch.questions.forEach((pq) => {
    const chapter = quiz.chapters[pq.chapterIndex];
    if (!chapter) {
      console.error(`Chapter ${pq.chapterIndex} not found`);
      return;
    }

    const question = chapter.questions[pq.questionIndex];
    if (!question) {
      console.error(`Question ${pq.questionIndex} not found in chapter ${pq.chapterIndex}`);
      return;
    }

    if (question.questionId !== pq.questionId) {
      console.error(`Question ID mismatch: expected ${pq.questionId}, got ${question.questionId}`);
      return;
    }

    // Update the question
    question.options = pq.options;
    question.explanationText = pq.explanationText;
    updatedCount++;
  });
});

console.log(`Updated ${updatedCount} questions`);

// Write the updated quiz
const outputPath = path.join(__dirname, '../public/default-quiz.json');
fs.writeFileSync(outputPath, JSON.stringify(quiz, null, 2));
console.log(`Written updated quiz to ${outputPath}`);

// Validate
const totalQuestions = quiz.chapters.reduce((sum, ch) => sum + ch.questions.length, 0);
console.log(`\nValidation:`);
console.log(`- Total chapters: ${quiz.chapters.length}`);
console.log(`- Total questions: ${totalQuestions}`);

// Check for any issues
let issues = 0;
quiz.chapters.forEach((chapter, ci) => {
  chapter.questions.forEach((q, qi) => {
    // Check correctOptionIds reference valid options
    const optionIds = q.options.map((o) => o.optionId);
    q.correctOptionIds.forEach((correctId) => {
      if (!optionIds.includes(correctId)) {
        console.error(`Invalid correctOptionId ${correctId} in ${q.questionId}`);
        issues++;
      }
    });

    // Check for duplicate option IDs
    const uniqueIds = new Set(optionIds);
    if (uniqueIds.size !== optionIds.length) {
      console.error(`Duplicate option IDs in ${q.questionId}`);
      issues++;
    }
  });
});

if (issues === 0) {
  console.log('- No validation issues found');
} else {
  console.log(`- Found ${issues} validation issues`);
}

File: scripts/merge-batches.py
#!/usr/bin/env python3
"""
Script to merge processed batch files back into the original quiz
"""

import json
import os
from pathlib import Path

# Get script directory
script_dir = Path(__file__).parent
project_root = script_dir.parent

# Read the original quiz
quiz_path = project_root / "public" / "default-quiz.json"
with open(quiz_path, 'r', encoding='utf-8') as f:
    quiz = json.load(f)

# Read all processed batch files
batch_dir = project_root / "temp-batches"
processed_dir = batch_dir / "processed"

if not processed_dir.exists():
    print("No processed directory found. Run subagents first.")
    exit(1)

processed_files = sorted(processed_dir.glob("*.json"))
print(f"Found {len(processed_files)} processed batch files")

updated_count = 0

for file_path in processed_files:
    with open(file_path, 'r', encoding='utf-8') as f:
        batch = json.load(f)
    
    for pq in batch['questions']:
        chapter_index = pq['chapterIndex']
        question_index = pq['questionIndex']
        
        if chapter_index >= len(quiz['chapters']):
            print(f"Chapter {chapter_index} not found")
            continue
        
        chapter = quiz['chapters'][chapter_index]
        
        if question_index >= len(chapter['questions']):
            print(f"Question {question_index} not found in chapter {chapter_index}")
            continue
        
        question = chapter['questions'][question_index]
        
        if question['questionId'] != pq['questionId']:
            print(f"Question ID mismatch: expected {pq['questionId']}, got {question['questionId']}")
            continue
        
        # Update the question
        question['options'] = pq['options']
        question['explanationText'] = pq['explanationText']
        updated_count += 1

print(f"Updated {updated_count} questions")

# Write the updated quiz
output_path = project_root / "public" / "default-quiz.json"
with open(output_path, 'w', encoding='utf-8') as f:
    json.dump(quiz, f, indent=2, ensure_ascii=False)
print(f"Written updated quiz to {output_path}")

# Validate
total_questions = sum(len(ch['questions']) for ch in quiz['chapters'])
print(f"\nValidation:")
print(f"- Total chapters: {len(quiz['chapters'])}")
print(f"- Total questions: {total_questions}")

# Check for any issues
issues = 0
for ci, chapter in enumerate(quiz['chapters']):
    for qi, q in enumerate(chapter['questions']):
        # Check correctOptionIds reference valid options
        option_ids = [o['optionId'] for o in q['options']]
        for correct_id in q['correctOptionIds']:
            if correct_id not in option_ids:
                print(f"Invalid correctOptionId {correct_id} in {q['questionId']}")
                issues += 1
        
        # Check for duplicate option IDs
        if len(set(option_ids)) != len(option_ids):
            print(f"Duplicate option IDs in {q['questionId']}")
            issues += 1

if issues == 0:
    print("- No validation issues found")
else:
    print(f"- Found {issues} validation issues")
